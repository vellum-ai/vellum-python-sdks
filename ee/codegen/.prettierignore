# Ignore snapshots
*.ts.snap

# Ignore generated assets
src/assets/node-definitions.json
