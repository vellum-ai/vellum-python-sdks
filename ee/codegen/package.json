{
  "name": "@vellum-ai/vellum-codegen",
  "version": "0.13.21",
  "main": "lib/src/index.js",
  "typings": "lib/src/index.d.ts",
  "files": [
    "README.md",
    "lib"
  ],
  "engines": {
    "node": ">=18.0.0"
  },
  "publishConfig": {
    "registry": "https://us-central1-npm.pkg.dev/vocify-prod/vellum-private-npm/"
  },
  "scripts": {
    "gar-login": "npx google-artifactregistry-auth",
    "clean": "rm -rf lib && rm -f tsconfig.tsbuildinfo",
    "build": "tsc && tsc-alias",
    "format": "prettier --write .",
    "format:check": "prettier --check . '!integration/**' '!lib/**'",
    "lint": "eslint .",
    "lint:fix": "eslint . --fix",
    "test": "vitest --run",
    "prepublishOnly": "npm run build && npm run gar-login",
    "test:update": "vitest --run -u",
    "test:debug": "vitest --run --inspect-brk --no-file-parallelism",
    "types": "tsc --noEmit",
    "start": "node --loader ts-node/esm --experimental-specifier-resolution=node src/index.ts",
    "version:codegen": "node scripts/version.js",
    "upgrade-codegen-service": "node --loader ts-node/esm --experimental-specifier-resolution=node  scripts/upgrade-codegen-service.mts"
  },
  "dependencies": {
    "@fern-api/python-ast": "^0.0.21",
    "@fern-fern/generator-cli-sdk": "file:./stubs/generator-cli-sdk",
    "@fern-fern/generator-exec-sdk": "file:./stubs/generator-exec-sdk",
    "lodash": "^4.17.21",
    "uuid": "^11.0.3",
    "vellum-ai": "0.10.7"
  },
  "devDependencies": {
    "@octokit/auth-app": "^7.1.3",
    "@types/lodash": "^4.17.13",
    "@types/node": "^22.10.2",
    "@typescript-eslint/eslint-plugin": "^8.9.0",
    "@typescript-eslint/parser": "^8.9.0",
    "eslint": "^8.56.0",
    "eslint-config-prettier": "^8.8.0",
    "eslint-import-resolver-typescript": "^3.5.5",
    "eslint-plugin-import": "^2.31.0",
    "eslint-plugin-no-relative-import-paths": "^1.5.2",
    "eslint-plugin-prettier": "^4.2.1",
    "eslint-plugin-unused-imports": "^4.1.4",
    "google-artifactregistry-auth": "3.2.0",
    "prettier": "^2.8.6",
    "ts-node": "^10.9.2",
    "tsc-alias": "1.8.10",
    "tsconfig-paths": "4.2.0",
    "typescript": "^5.6.3",
    "vitest": "^2.1.3"
  },
  "overrides": {
    "@fern-api/logger": "^0.4.24-rc1"
  }
}
