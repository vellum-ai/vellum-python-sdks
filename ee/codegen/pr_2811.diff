diff --git a/ee/codegen_integration/fixtures/simple_manual_trigger_workflow/display_data/simple_manual_trigger_workflow.json b/ee/codegen_integration/fixtures/simple_manual_trigger_workflow/display_data/simple_manual_trigger_workflow.json
index 297c5caac..ae2ba9b41 100644
--- a/ee/codegen_integration/fixtures/simple_manual_trigger_workflow/display_data/simple_manual_trigger_workflow.json
+++ b/ee/codegen_integration/fixtures/simple_manual_trigger_workflow/display_data/simple_manual_trigger_workflow.json
@@ -2,7 +2,7 @@
   "workflow_raw_data": {
     "nodes": [
       {
-        "id": "63884a7b-c01c-4cbc-b8d4-abe0a8796f6b",
+        "id": "b09c1902-3cca-4c79-b775-4c32e3e88466",
         "type": "ENTRYPOINT",
         "inputs": [],
         "data": {
@@ -147,7 +147,7 @@
     "edges": [
       {
         "id": "9eaed9b1-5803-4bc4-bdba-b614f4f465b5",
-        "source_node_id": "63884a7b-c01c-4cbc-b8d4-abe0a8796f6b",
+        "source_node_id": "b09c1902-3cca-4c79-b775-4c32e3e88466",
         "source_handle_id": "eba8fd73-57ab-4d7b-8f75-b54dbe5fc8ba",
         "target_node_id": "3669684e-500e-4034-be4a-7ee70824b272",
         "target_handle_id": "7da60409-5790-43fa-80b2-ad88da2812df",
diff --git a/ee/vellum_ee/workflows/display/workflows/base_workflow_display.py b/ee/vellum_ee/workflows/display/workflows/base_workflow_display.py
index 74d033afe..c63d7c2e0 100644
--- a/ee/vellum_ee/workflows/display/workflows/base_workflow_display.py
+++ b/ee/vellum_ee/workflows/display/workflows/base_workflow_display.py
@@ -179,9 +179,22 @@ def serialize(self) -> JsonObject:
         serialized_nodes: Dict[UUID, JsonObject] = {}
         edges: JsonArray = []

-        # Add a single synthetic node for the workflow entrypoint
-        entrypoint_node_id = self.display_context.workflow_display.entrypoint_node_id
+        # Get all trigger edges from the workflow's subgraphs to check if trigger exists
+        trigger_edges = []
+        for subgraph in self._workflow.get_subgraphs():
+            trigger_edges.extend(list(subgraph.trigger_edges))
+
+        # Determine entrypoint node ID: use trigger ID if trigger exists, otherwise use default
+        # This maintains backwards compatibility while linking trigger and entrypoint
+        if len(trigger_edges) > 0:
+            trigger_class = trigger_edges[0].trigger_class
+            entrypoint_node_id = get_trigger_id(trigger_class)
+        else:
+            entrypoint_node_id = self.display_context.workflow_display.entrypoint_node_id
+
         entrypoint_node_source_handle_id = self.display_context.workflow_display.entrypoint_node_source_handle_id
+
+        # Always add entrypoint node for backwards compatibility
         serialized_nodes[entrypoint_node_id] = {
             "id": str(entrypoint_node_id),
             "type": "ENTRYPOINT",
@@ -340,7 +353,8 @@ def serialize(self) -> JsonObject:
                 ValueError("Unable to serialize terminal nodes that are not referenced by workflow outputs.")
             )

-        # Add an edge for each edge in the workflow
+        # Add edges from entrypoint node to first nodes
+        # These edges always exist for backwards compatibility
         for target_node, entrypoint_display in self.display_context.entrypoint_displays.items():
             unadorned_target_node = get_unadorned_node(target_node)
             # Skip edges to invalid nodes
