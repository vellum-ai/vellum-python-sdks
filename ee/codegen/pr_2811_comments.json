[{"url":"https://api.github.com/repos/vellum-ai/vellum-python-sdks/pulls/comments/2446296700","pull_request_review_id":3358153351,"id":2446296700,"node_id":"PRRC_kwDOJTq8L86Rz4Z8","diff_hunk":"@@ -2,7 +2,7 @@\n   \"workflow_raw_data\": {\n     \"nodes\": [\n       {\n-        \"id\": \"63884a7b-c01c-4cbc-b8d4-abe0a8796f6b\",\n+        \"id\": \"b09c1902-3cca-4c79-b775-4c32e3e88466\",","path":"ee/codegen_integration/fixtures/simple_manual_trigger_workflow/display_data/simple_manual_trigger_workflow.json","commit_id":"56a21972e55f3b3e00f760d8d0f5243fb4dbca81","original_commit_id":"56a21972e55f3b3e00f760d8d0f5243fb4dbca81","user":{"login":"alex-nork","id":48630278,"node_id":"MDQ6VXNlcjQ4NjMwMjc4","avatar_url":"https://avatars.githubusercontent.com/u/48630278?v=4","gravatar_id":"","url":"https://api.github.com/users/alex-nork","html_url":"https://github.com/alex-nork","followers_url":"https://api.github.com/users/alex-nork/followers","following_url":"https://api.github.com/users/alex-nork/following{/other_user}","gists_url":"https://api.github.com/users/alex-nork/gists{/gist_id}","starred_url":"https://api.github.com/users/alex-nork/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/alex-nork/subscriptions","organizations_url":"https://api.github.com/users/alex-nork/orgs","repos_url":"https://api.github.com/users/alex-nork/repos","events_url":"https://api.github.com/users/alex-nork/events{/privacy}","received_events_url":"https://api.github.com/users/alex-nork/received_events","type":"User","user_view_type":"public","site_admin":false},"body":"this matches the manual trigger's `id` below","created_at":"2025-10-20T22:46:25Z","updated_at":"2025-10-20T22:46:25Z","html_url":"https://github.com/vellum-ai/vellum-python-sdks/pull/2811#discussion_r2446296700","pull_request_url":"https://api.github.com/repos/vellum-ai/vellum-python-sdks/pulls/2811","_links":{"self":{"href":"https://api.github.com/repos/vellum-ai/vellum-python-sdks/pulls/comments/2446296700"},"html":{"href":"https://github.com/vellum-ai/vellum-python-sdks/pull/2811#discussion_r2446296700"},"pull_request":{"href":"https://api.github.com/repos/vellum-ai/vellum-python-sdks/pulls/2811"}},"reactions":{"url":"https://api.github.com/repos/vellum-ai/vellum-python-sdks/pulls/comments/2446296700/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"start_line":null,"original_start_line":null,"start_side":null,"line":5,"original_line":5,"side":"RIGHT","author_association":"MEMBER","original_position":5,"position":5,"subject_type":"line"},{"url":"https://api.github.com/repos/vellum-ai/vellum-python-sdks/pulls/comments/2449261312","pull_request_review_id":3362073135,"id":2449261312,"node_id":"PRRC_kwDOJTq8L86R_MMA","diff_hunk":"@@ -179,9 +179,22 @@ def serialize(self) -> JsonObject:\n         serialized_nodes: Dict[UUID, JsonObject] = {}\n         edges: JsonArray = []\n \n-        # Add a single synthetic node for the workflow entrypoint\n-        entrypoint_node_id = self.display_context.workflow_display.entrypoint_node_id\n+        # Get all trigger edges from the workflow's subgraphs to check if trigger exists\n+        trigger_edges = []\n+        for subgraph in self._workflow.get_subgraphs():\n+            trigger_edges.extend(list(subgraph.trigger_edges))\n+\n+        # Determine entrypoint node ID: use trigger ID if trigger exists, otherwise use default\n+        # This maintains backwards compatibility while linking trigger and entrypoint\n+        if len(trigger_edges) > 0:\n+            trigger_class = trigger_edges[0].trigger_class","path":"ee/vellum_ee/workflows/display/workflows/base_workflow_display.py","commit_id":"56a21972e55f3b3e00f760d8d0f5243fb4dbca81","original_commit_id":"56a21972e55f3b3e00f760d8d0f5243fb4dbca81","user":{"login":"dvargas92495","id":7143571,"node_id":"MDQ6VXNlcjcxNDM1NzE=","avatar_url":"https://avatars.githubusercontent.com/u/7143571?v=4","gravatar_id":"","url":"https://api.github.com/users/dvargas92495","html_url":"https://github.com/dvargas92495","followers_url":"https://api.github.com/users/dvargas92495/followers","following_url":"https://api.github.com/users/dvargas92495/following{/other_user}","gists_url":"https://api.github.com/users/dvargas92495/gists{/gist_id}","starred_url":"https://api.github.com/users/dvargas92495/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/dvargas92495/subscriptions","organizations_url":"https://api.github.com/users/dvargas92495/orgs","repos_url":"https://api.github.com/users/dvargas92495/repos","events_url":"https://api.github.com/users/dvargas92495/events{/privacy}","received_events_url":"https://api.github.com/users/dvargas92495/received_events","type":"User","user_view_type":"public","site_admin":false},"body":"[blocking] instead of `trigger_edges[0]`, I would expect to align it with only the `ManualTrigger`","created_at":"2025-10-21T18:07:54Z","updated_at":"2025-10-21T18:08:35Z","html_url":"https://github.com/vellum-ai/vellum-python-sdks/pull/2811#discussion_r2449261312","pull_request_url":"https://api.github.com/repos/vellum-ai/vellum-python-sdks/pulls/2811","_links":{"self":{"href":"https://api.github.com/repos/vellum-ai/vellum-python-sdks/pulls/comments/2449261312"},"html":{"href":"https://github.com/vellum-ai/vellum-python-sdks/pull/2811#discussion_r2449261312"},"pull_request":{"href":"https://api.github.com/repos/vellum-ai/vellum-python-sdks/pulls/2811"}},"reactions":{"url":"https://api.github.com/repos/vellum-ai/vellum-python-sdks/pulls/comments/2449261312/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"start_line":null,"original_start_line":null,"start_side":null,"line":190,"original_line":190,"side":"RIGHT","author_association":"MEMBER","original_position":14,"position":14,"subject_type":"line"},{"url":"https://api.github.com/repos/vellum-ai/vellum-python-sdks/pulls/comments/2449262487","pull_request_review_id":3362073135,"id":2449262487,"node_id":"PRRC_kwDOJTq8L86R_MeX","diff_hunk":"@@ -179,9 +179,22 @@ def serialize(self) -> JsonObject:\n         serialized_nodes: Dict[UUID, JsonObject] = {}\n         edges: JsonArray = []\n \n-        # Add a single synthetic node for the workflow entrypoint\n-        entrypoint_node_id = self.display_context.workflow_display.entrypoint_node_id\n+        # Get all trigger edges from the workflow's subgraphs to check if trigger exists\n+        trigger_edges = []\n+        for subgraph in self._workflow.get_subgraphs():\n+            trigger_edges.extend(list(subgraph.trigger_edges))","path":"ee/vellum_ee/workflows/display/workflows/base_workflow_display.py","commit_id":"56a21972e55f3b3e00f760d8d0f5243fb4dbca81","original_commit_id":"56a21972e55f3b3e00f760d8d0f5243fb4dbca81","user":{"login":"dvargas92495","id":7143571,"node_id":"MDQ6VXNlcjcxNDM1NzE=","avatar_url":"https://avatars.githubusercontent.com/u/7143571?v=4","gravatar_id":"","url":"https://api.github.com/users/dvargas92495","html_url":"https://github.com/dvargas92495","followers_url":"https://api.github.com/users/dvargas92495/followers","following_url":"https://api.github.com/users/dvargas92495/following{/other_user}","gists_url":"https://api.github.com/users/dvargas92495/gists{/gist_id}","starred_url":"https://api.github.com/users/dvargas92495/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/dvargas92495/subscriptions","organizations_url":"https://api.github.com/users/dvargas92495/orgs","repos_url":"https://api.github.com/users/dvargas92495/repos","events_url":"https://api.github.com/users/dvargas92495/events{/privacy}","received_events_url":"https://api.github.com/users/dvargas92495/received_events","type":"User","user_view_type":"public","site_admin":false},"body":"[oos] how are we planning to serialize trigger edges?","created_at":"2025-10-21T18:08:16Z","updated_at":"2025-10-21T18:08:35Z","html_url":"https://github.com/vellum-ai/vellum-python-sdks/pull/2811#discussion_r2449262487","pull_request_url":"https://api.github.com/repos/vellum-ai/vellum-python-sdks/pulls/2811","_links":{"self":{"href":"https://api.github.com/repos/vellum-ai/vellum-python-sdks/pulls/comments/2449262487"},"html":{"href":"https://github.com/vellum-ai/vellum-python-sdks/pull/2811#discussion_r2449262487"},"pull_request":{"href":"https://api.github.com/repos/vellum-ai/vellum-python-sdks/pulls/2811"}},"reactions":{"url":"https://api.github.com/repos/vellum-ai/vellum-python-sdks/pulls/comments/2449262487/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"start_line":null,"original_start_line":null,"start_side":null,"line":185,"original_line":185,"side":"RIGHT","author_association":"MEMBER","original_position":9,"position":9,"subject_type":"line"}]
