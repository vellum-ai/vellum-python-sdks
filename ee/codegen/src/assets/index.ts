// This directory contains generated assets that are copied during postinstall
// and included in the npm package build. The node-definitions.json file is
// generated from the Python SDK and copied here for use in code generation.
