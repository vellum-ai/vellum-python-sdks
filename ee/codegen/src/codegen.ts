import {
  ErrorLogFile,
  InitFile,
  Inputs,
  NodeInput,
  VellumValue,
  VellumVariable,
  Workflow,
} from "./generators";
import { BasePersistedFile } from "./generators/base-persisted-file";
import { State } from "./generators/state";
import { WorkflowSandboxFile } from "./generators/workflow-sandbox-file";

export function vellumVariable(args: VellumVariable.Args): VellumVariable {
  return new VellumVariable(args);
}

export function vellumValue(args: VellumValue.Args): VellumValue {
  return new VellumValue(args);
}

export function nodeInput(args: NodeInput.Args): NodeInput {
  return new NodeInput(args);
}

export function inputs(args: Inputs.Args): Inputs {
  return new Inputs(args);
}

export function state(args: State.Args): State {
  return new State(args);
}

export function workflow(args: Workflow.Args): Workflow {
  return new Workflow(args);
}

export function initFile(args: InitFile.Args): InitFile {
  return new InitFile(args);
}

export function errorLogFile(args: BasePersistedFile.Args): ErrorLogFile {
  return new ErrorLogFile(args);
}

export function workflowSandboxFile(
  args: WorkflowSandboxFile.Args
): WorkflowSandboxFile {
  return new WorkflowSandboxFile(args);
}
