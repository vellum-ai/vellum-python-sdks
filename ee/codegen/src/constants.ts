/* Module paths */
export const VELLUM_CLIENT_MODULE_PATH = ["vellum"] as const;
export const VELLUM_WORKFLOW_NODE_BASE_TYPES_PATH = [
  "vellum",
  "workflows",
  "nodes",
  "displayable",
  "bases",
  "types",
] as const;
export const VELLUM_WORKFLOW_GRAPH_MODULE_PATH = [
  "vellum",
  "workflows",
  "graph",
] as const;
export const VELLUM_WORKFLOW_NODES_MODULE_PATH = [
  "vellum",
  "workflows",
  "nodes",
] as const;
export const VELLUM_WORKFLOW_CONSTANTS_PATH = [
  "vellum",
  "workflows",
  "references",
  "constant",
] as const;
export const VELLUM_WORKFLOWS_DISPLAY_BASE_PATH = [
  "vellum_ee",
  "workflows",
  "display",
  "base",
] as const;
export const VELLUM_WORKFLOW_EDITOR_TYPES_PATH = [
  "vellum_ee",
  "workflows",
  "display",
  "editor",
] as const;
/* Class names */
export const OUTPUTS_CLASS_NAME = "Outputs";
export const PORTS_CLASS_NAME = "Ports";
export const DEFAULT_PORT_NAME = "default";

/* File names */
export const INIT_FILE_NAME = "__init__.py";

export const GENERATED_WORKFLOW_MODULE_NAME = "workflow";

export const GENERATED_INPUTS_MODULE_NAME = "inputs";

export const GENERATED_NODES_MODULE_NAME = "nodes";

export const GENERATED_DISPLAY_MODULE_NAME = "display";
export const GENERATED_NODES_PATH = [GENERATED_NODES_MODULE_NAME] as const;

export const GENERATED_DISPLAY_NODE_MODULE_PATH = [
  GENERATED_DISPLAY_MODULE_NAME,
  GENERATED_NODES_MODULE_NAME,
] as const;

export const GENERATED_NESTED_NODE_MODULE_NAME = "node";
