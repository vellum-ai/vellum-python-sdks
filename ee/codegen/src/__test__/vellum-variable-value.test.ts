import { Writer } from "@fern-api/python-ast/core/Writer";

import * as codegen from "src/codegen";

describe("VellumValue", () => {
  let writer: Writer;

  beforeEach(() => {
    writer = new Writer();
  });

  describe("STRING", () => {
    it("should write a STRING value correctly", async () => {
      const stringValue = codegen.vellumValue({
        vellumValue: {
          type: "STRING",
          value: "Hello, World!",
        },
      });
      stringValue.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
      expect(stringValue.getReferences()).toHaveLength(0);
    });
  });

  describe("NUMBER", () => {
    it("should write a NUMBER value correctly", async () => {
      const numberValue = codegen.vellumValue({
        vellumValue: {
          type: "NUMBER",
          value: 42,
        },
      });
      numberValue.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
      expect(numberValue.getReferences()).toHaveLength(0);
    });
  });

  describe("CHAT_HISTORY", () => {
    it("should write a CHAT_HISTORY value with just text", async () => {
      const chatHistoryValue = codegen.vellumValue({
        vellumValue: {
          type: "CHAT_HISTORY",
          value: [
            {
              role: "USER",
              text: "Hello, AI!",
            },
          ],
        },
      });
      chatHistoryValue.write(writer);
      expect(await writer.toString()).toMatchSnapshot();
      expect(chatHistoryValue.getReferences()).toHaveLength(1);
    });

    it("should write a CHAT_HISTORY value with a string content", async () => {
      const chatHistoryValue = codegen.vellumValue({
        vellumValue: {
          type: "CHAT_HISTORY",
          value: [
            {
              role: "USER",
              content: { type: "STRING", value: "Hello, AI!" },
            },
          ],
        },
      });
      chatHistoryValue.write(writer);
      expect(await writer.toString()).toMatchSnapshot();
      expect(chatHistoryValue.getReferences()).toHaveLength(2);
    });

    it("should write nothing when CHAT_HISTORY value is null", async () => {
      const chatHistoryValue = codegen.vellumValue({
        vellumValue: {
          type: "CHAT_HISTORY",
        },
      });
      chatHistoryValue.write(writer);
      expect(await writer.toString()).toMatchSnapshot();
      expect(chatHistoryValue.getReferences()).toHaveLength(0);
    });

    it("should write empty list when CHAT_HISTORY value is empty list", async () => {
      const chatHistoryValue = codegen.vellumValue({
        vellumValue: {
          type: "CHAT_HISTORY",
          value: [],
        },
      });
      chatHistoryValue.write(writer);
      expect(await writer.toString()).toMatchSnapshot();
      expect(chatHistoryValue.getReferences()).toHaveLength(0);
    });
  });

  describe("JSON", () => {
    it("should write a JSON value correctly", async () => {
      const jsonValue = codegen.vellumValue({
        vellumValue: {
          type: "JSON",
          value: {
            key: "value",
            nested: { array: [1, 2, 3] },
          },
        },
      });
      jsonValue.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
      expect(jsonValue.getReferences()).toHaveLength(0);
    });
  });

  describe("ERROR", () => {
    it("should write a ERROR value correctly", async () => {
      const errorValue = codegen.vellumValue({
        vellumValue: {
          type: "ERROR",
          value: {
            message: "This is an error!",
            code: "INTERNAL_SERVER_ERROR",
          },
        },
      });
      errorValue.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
      expect(errorValue.getReferences()).toHaveLength(1);
    });
  });

  describe("IMAGE", () => {
    it("should write a IMAGE value correctly", async () => {
      const imageValue = codegen.vellumValue({
        vellumValue: {
          type: "IMAGE",
          value: {
            src: "https://example.com/image.png",
          },
        },
      });
      imageValue.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
      expect(imageValue.getReferences()).toHaveLength(1);
    });
  });

  describe("FUNCTION_CALL", () => {
    it("should write a FUNCTION_CALL value correctly", async () => {
      const functionCallValue = codegen.vellumValue({
        vellumValue: {
          type: "FUNCTION_CALL",
          value: {
            arguments: {
              key: "value",
            },
            name: "test",
            id: "123",
          },
        },
      });
      functionCallValue.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
      expect(functionCallValue.getReferences()).toHaveLength(1);
    });
  });

  describe("ARRAY", () => {
    it("should write a ARRAY value correctly", async () => {
      const arrayValue = codegen.vellumValue({
        vellumValue: {
          type: "ARRAY",
          value: [
            {
              type: "STRING",
              value: "Hello, World!",
            },
            {
              type: "NUMBER",
              value: 42,
            },
            {
              type: "AUDIO",
              value: {
                src: "https://example.com/audio.mp3",
              },
            },
          ],
        },
      });
      arrayValue.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
      expect(arrayValue.getReferences()).toHaveLength(1);
    });
  });

  describe("AUDIO", () => {
    it("should write a AUDIO value correctly", async () => {
      const audioValue = codegen.vellumValue({
        vellumValue: {
          type: "AUDIO",
          value: {
            src: "https://example.com/audio.mp3",
          },
        },
      });
      audioValue.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
      expect(audioValue.getReferences()).toHaveLength(1);
    });
  });

  describe("SEARCH_RESULTS", () => {
    it("should write a SEARCH_RESULTS value correctly", async () => {
      const searchResultsValue = codegen.vellumValue({
        vellumValue: {
          type: "SEARCH_RESULTS",
          value: [
            {
              text: "Hello, World!",
              score: 0.5,
              keywords: ["hello", "world"],
              document: {
                id: "123",
                label: "Example Document",
              },
            },
            {
              text: "Hello, AI!",
              score: 0.7,
              keywords: ["hello", "ai"],
              document: {
                id: "456",
                label: "Another Document",
              },
            },
          ],
        },
      });
      searchResultsValue.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
      expect(searchResultsValue.getReferences()).toHaveLength(4);
    });
  });
});
