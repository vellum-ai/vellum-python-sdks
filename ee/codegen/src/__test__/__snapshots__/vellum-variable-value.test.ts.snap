// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VellumValue > ARRAY > should write a ARRAY value correctly 1`] = `
"[
    "Hello, World!",
    42,
    VellumAudio(src="https://example.com/audio.mp3"),
]
"
`;

exports[`VellumValue > AUDIO > should write a AUDIO value correctly 1`] = `
"VellumAudio(src="https://example.com/audio.mp3")
"
`;

exports[`VellumValue > CHAT_HISTORY > should write a CHAT_HISTORY value with a string content 1`] = `"[ChatMessage(role="USER", content=StringChatMessageContent(value="Hello, AI!")),]"`;

exports[`VellumValue > CHAT_HISTORY > should write a CHAT_HISTORY value with just text 1`] = `"[ChatMessage(role="USER", text="Hello, AI!"),]"`;

exports[`VellumValue > CHAT_HISTORY > should write empty list when CHAT_HISTORY value is empty list 1`] = `"[]"`;

exports[`VellumValue > CHAT_HISTORY > should write nothing when CHAT_HISTORY value is null 1`] = `"None"`;

exports[`VellumValue > ERROR > should write a ERROR value correctly 1`] = `
"VellumError(message="This is an error!", code="INTERNAL_SERVER_ERROR")
"
`;

exports[`VellumValue > FUNCTION_CALL > should write a FUNCTION_CALL value correctly 1`] = `
"FunctionCall(
    arguments={
        "key": "value",
    },
    name="test",
    id="123",
)
"
`;

exports[`VellumValue > IMAGE > should write a IMAGE value correctly 1`] = `
"VellumImage(src="https://example.com/image.png")
"
`;

exports[`VellumValue > JSON > should write a JSON value correctly 1`] = `
"{
    "key": "value",
    "nested": {
        "array": [
            1,
            2,
            3,
        ],
    },
}
"
`;

exports[`VellumValue > NUMBER > should write a NUMBER value correctly 1`] = `
"42
"
`;

exports[`VellumValue > SEARCH_RESULTS > should write a SEARCH_RESULTS value correctly 1`] = `
"[
    SearchResult(
        text="Hello, World!",
        score=0.5,
        keywords=["hello", "world"],
        document=Document(id="123", label="Example Document"),
    ),
    SearchResult(
        text="Hello, AI!",
        score=0.7,
        keywords=["hello", "ai"],
        document=Document(id="456", label="Another Document"),
    ),
]
"
`;

exports[`VellumValue > STRING > should write a STRING value correctly 1`] = `
""Hello, World!"
"
`;
